src/pd/leader.rs

```diff
@@ -1,51 +1,222 @@
+                    Err(()) => Err(Error::Other(box_err!("failed to recv tso requests"))),
+                })).then(|r| match r {
+                    Ok((mut sender, _)) => {
+                        sender.get_mut().cancel();
+                        Ok(())
+                    }
+                    Err(e) => {
+                        error!("failed to send tso requests: {:?}", e);
+                        Err(())
+                    }
+                }),
+        );
+        handle.spawn(
+            rx.for_each(move |resp| {
+                let mut client = client.wl();
+                let reactor = &mut client.reactor;
+                let tso_pending = reactor.tso_pending.take().unwrap();
+                reactor.schedule(PdTask::TsoResponse(tso_pending, resp));
+                if !reactor.tso_batch.is_empty() {
+                    /* schedule another tso_batch of request */
+                    reactor.schedule(PdTask::TsoRequest);
+                }
+                Ok(())
+            }).map_err(|e| panic!("unexpected error: {:?}", e)),
+        );
+    }
+
+    fn tso_request(client: &Arc<RwLock<LeaderClient>>) {
+        let mut client = client.wl();
+        let cluster_id = client.cluster_id;
+        let reactor = &mut client.reactor;
+        let mut tso_batch = reactor.tso_buffer.take().unwrap();
+        tso_batch.extend(reactor.tso_batch.drain(..));
+        let mut request = request!(cluster_id, TsoRequest);
+        request.set_count(tso_batch.len() as u32);
+        reactor.tso_pending = Some(tso_batch);
+        reactor.tso_tx.unbounded_send(request).unwrap();
+    }
+
+    fn tso_response(
+        client: &Arc<RwLock<LeaderClient>>,
+        mut requests: Vec<TsoChannel>,
+        response: &TsoResponse,
+    ) {
+        let timestamp = response.get_timestamp();
+        for (offset, request) in requests.drain(..).enumerate() {
+            request
+                .send(PdTimestamp {
+                    physical: timestamp.physical,
+                    logical: timestamp.logical + offset as i64,
+                }).unwrap();
+        }
+        client.wl().reactor.tso_buffer = Some(requests);
+    }
+
+    fn dispatch(client: &Arc<RwLock<LeaderClient>>, task: PdTask, handle: &OtherHandle) {
+        match task {
+            PdTask::TsoRequest => Self::tso_request(client),
+            PdTask::TsoResponse(requests, response) => Self::tso_response(client, requests, &response),
+            PdTask::TsoInit => Self::init(client, handle),
+            PdTask::Request(task) => handle.spawn(task),
+        }
+    }
+
+    fn get_ts(&mut self) -> PdFuture<PdTimestamp> {
+        let timer = Instant::now();
+        let (tx, rx) = oneshot::channel::<PdTimestamp>();
+        self.tso_batch.push(tx);
+        if self.tso_pending.is_none() {
+            /* schedule tso request to run */
+            self.schedule(PdTask::TsoRequest);
+        }
+        Box::new(rx.map_err(Error::Canceled).and_then(move |ts| {
```

Comment:
You can maybe use `impl Future` as a return type and avoid the box. Reference: https://tokio.rs/docs/going-deeper/returning/

Reply:
Hm, while I changing it to `impl Future` I found I eventually still had to add Box. Since we are using Trait to define PdClient interface and `impl Trait` isn't possible for trait, we still have to use boxed version here.

So shell we have another discussion about using trait here. Looks like testability is the major benefit of using trait, do we want to pay the extra cost for this. Or maybe we can change the trait to expose only synchronous version of methods and leave future methods in the concrete implementation.

Reply:
@Hoverbear I pushed another commit to move all Future related interfaces to struct and keep synchronous version in trait. It looks not too bad to me.

Reply:
> impl Trait isn't possible for trait

Ohhh, so interesting. It is sad `impl trait` is so limited. :(

I think using `impl trait` without the `Box::new()` can help save an allocation, so I try to favor it where possible.

I think my favorite use is in argument position though for accepting iterables and optionals.

```rust
use std::collections::HashSet;

fn main() {
    println!("{:#?}", unique(vec![123, 1234, 123]).collect::<Vec<u64>>());
    optional_arg(None);
    optional_arg(Some(1));
    optional_arg(1);
}

fn unique<'a>(v: impl IntoIterator<Item=u64>) -> impl Iterator<Item=u64> {
    let v = v.into_iter().collect::<HashSet<_>>();
    v.into_iter()
}

fn optional_arg(v: impl Into<Option<u64>>) {
    let v = v.into().unwrap_or_else(Default::default);
}
```

Reply:
Just come up with another idea. To avoid confusing people with identical method name but different return type in trait and struct. We can expose asynchronous interface as different name with boxed suffix for example however return a Boxed future instead. 
How does that sound?

Reply:
I think that is a bit of a misnomer. Since `Box<T> impl Deref<T>` something being boxed is not terribly significant. Whereas something being a `Future` requires different handling than something being sync.

I see some valid use cases for having both async/sync calls available... But I also don't see a problem with returning futures and letting users wait to have them resolved if they need sync.

Perhaps it was better at first async as the only option, and just return boxes. 😢

Reply:
Hm, are we considering adding sync call later on. If so I prefer keep async suffix in method name just like generated gRPC code. Otherwise we can just keep using the method name defined in proto files.

Reply:
I think if someone wants sync they can just `wait()` for the values perhaps.

Reply:
Got it.

Reply:
@Hoverbear Updated. Please take another look.

Reply:
Yes! I think this is the best. :) I am sorry for asking you to change something and it being misleading.